### DISCUSSION TITLE
How to use Riverpod in a pure Dart project?

### DISCUSSION THREAD

**Discussion by KristijanZic** [rrousselGit/riverpod#3694] [Q&A] [answered]
I have a pure Dart project where I'd like to use riverpod for dependency injection etc but I can't find any docs or examples for doing that. What would be the best approach? Using `ProviderContainer();` everywhere?

**Comment by snapsl** [comment]
> What would be the best approach?

I would recommend using `get_it` / `injectable`. Riverpod is probably not suitable for this purpose.

> Using ProviderContainer(); everywhere?

Yes, you need some container to access your providers.

**Reply by rrousselGit** [reply]
Riverpod is definitely suitable for pure Dart projects. I've done it multiple times

**Reply by snapsl** [reply]
In general you can use riverpod for pure dart projects, but it was specifically requested for DI. (#3501 "Riverpod is not about DI")

**Reply by rrousselGit** [reply]
It depends on what that "etc" means :)
I doubt they want to use Riverpod for _just_ DI

**Reply by KristijanZic** [reply]
I'm making a cli app that connects to pocket base. I need to provide the pocketbase service to the rest of the app and then have repositories that return stuff like AuthUser, various records that get fetched from the pocketbase trough the service etc.

**Reply by snapsl** [reply]
I don't see where you need the reactive component of riverpod here, such as listening to state changes, which is limited to the container.  For simple DI I would recommend the other packages.

**Reply by KristijanZic** [reply]
I have a case where I'd potentially have to react to isLoggedIn state change, or manually refresh the auth token periodically since the pocketbase doesn't support a refresh token etc.
But yeah, I think I could use just getIt for most of the stuff, I just wanted to see if I could use Riverpod in a Dart project and how to do it properly.

**Comment by rrousselGit** [comment, marked as answer]
Generally you'll want to create only **one** ProviderContainer, inside the `main` (not as a global!).
Then have the main `container.listen` some providers.

**Reply by KristijanZic** [reply]
So the approach is to create a container in the main and then pass it down everywhere? What if I used both getIt and riverpod and did something like: `getIt.registerSingleton<ProviderContainer>(container);`?

**Reply by rrousselGit** [reply]
There's not really a reason to do so.
